linkerd/app/core/src/control.rs

```diff
@@ -145,160 +145,376 @@ pub mod add_origin {
+    use tracing::debug;
 
-    #[derive(Clone, Debug)]
-    pub struct Layer {
-        dns: dns::Resolver,
+    #[derive(Clone)]
+    pub struct Resolve<R> {
+        dns: R,
     }
 
-    #[derive(Clone, Debug)]
-    pub struct Resolve<M> {
-        dns: dns::Resolver,
-        inner: M,
-    }
+    // === impl ControlPlaneResolve ===
 
-    #[pin_project]
-    pub struct Init<M>
-    where
-        M: tower::Service<client::Target>,
-    {
-        #[pin]
-        state: State<M>,
+    impl<R> Resolve<R> {
+        pub fn new(dns: R) -> Self {
+            Self { dns }
+        }
     }
 
-    #[pin_project(project = StateProj)]
-    enum State<M>
+    type UpdatesStream =
+        Pin<Box<dyn Stream<Item = Result<Update<Target>, Error>> + Send + 'static>>;
+
+    impl<R> tower::Service<ControlAddr> for Resolve<R>
     where
-        M: tower::Service<client::Target>,
+        R: dns::Resolver + Send + Sync + 'static,
     {
-        Resolve(#[pin] dns::IpAddrFuture, Option<(M, ControlAddr)>),
-        NotReady(M, Option<(SocketAddr, ControlAddr)>),
-        Inner(#[pin] M::Future),
-    }
-
-    #[derive(Debug)]
-    pub enum Error<I> {
-        Dns(dns::Error),
-        Inner(I),
-    }
+        type Response = resolve::FromStream<UpdatesStream>;
+        type Error = Error;
+        type Future = futures::future::Ready<Result<Self::Response, Self::Error>>;
 
-    // === impl Layer ===
+        fn poll_ready(&mut self, _: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+            Poll::Ready(Ok(()))
+        }
 
-    pub fn layer<M>(dns: dns::Resolver) -> impl svc::Layer<M, Service = Resolve<M>> + Clone
-    where
-        M: tower::Service<client::Target> + Clone,
-    {
-        svc::layer::mk(move |inner| Resolve {
-            dns: dns.clone(),
-            inner,
-        })
+        fn call(&mut self, target: ControlAddr) -> Self::Future {
+            futures::future::ok(resolve::from_stream::<UpdatesStream>(Box::pin(
+                resolution_stream(self.dns.clone(), target.clone()),
+            )))
+        }
     }
 
-    // === impl Resolve ===
+    // === impl ControlPlaneResolution ===
 
-    impl<M> tower::Service<ControlAddr> for Resolve<M>
-    where
-        M: tower::Service<client::Target> + Clone,
-    {
-        type Response = M::Response;
-        type Error = <Init<M> as TryFuture>::Error;
-        type Future = Init<M>;
+    pub fn diff_targets(
+        old_targets: &Vec<Target>,
+        new_targets: &Vec<Target>,
+    ) -> (Vec<(SocketAddr, Target)>, Vec<SocketAddr>) {
+        let mut adds = Vec::default();
+        let mut removes = Vec::default();
 
-        fn poll_ready(&mut self, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
-            self.inner.poll_ready(cx).map_err(Error::Inner)
+        for new_target in new_targets.iter() {
+            if !old_targets.contains(new_target) {
+                adds.push((new_target.addr, new_target.clone()))
+            }
         }
```

Comment:
Is it worth using a Set here? This messes up the order in which these endpoints are returned and causes problems. Does the .contains in a `Vec` pose a significant problem when considering the number of endpoints should not be so high (I suppose)?

Reply:
Sorry, I was not clear. This does not break anything. Having consistent order simply makes testing easier.